docs/server/source/drivers-clients/websocket-event-stream-api.rst

```diff
@@ -0,0 +1,105 @@
+
+BigchainDB provides real-time event streams over the WebSockets protocol with
+the Events Stream API.
+
+Connecting to a Events Stream from your application enables a BigchainDB node
+to notify you as events are processed, such as new `validated transactions <#valid-transactions>`_.
+
+
+Determining Support for the Events Stream API
+---------------------------------------------
+
+In practice, it's a good idea to make sure that the node you're connecting with
+has support for the Events Stream API. To do so, send a HTTP GET request to the
+node's :ref:`Root URL <bigchaindb-root-url>` and check that the response
+contains a ``streams_<version>`` property in ``_links``::
+
+    {
+      "_links": {
+         "streams_v1": "ws://example.com:9984/api/v1/streams/"
+      }
+    }
+
+
+Connection Keep Alive
+~~~~~~~~~~~~~~~~~~~~~
+
+The Events Stream API requires clients to signal that they'd like their
+connection to stay open by sending "pings" across the open connection.
+BigchainDB nodes will automatically close any connections that haven't sent a
+ping in the last three minutes.
+
+.. note::
+
+    While three minutes is the limit before a BigchainDB server will terminate
+    a connection, we suggest sending a ping every 30 seconds for better
+    reliability.
+
+A "ping" consists of a message containing only the string ``"ping"``, for example
+in JavaScript:
+
+.. code-block:: javascript
+
+    new WebSocket("...").send("ping")
+
+If the BigchainDB node received the ping, it'll respond back with a message
+containing only the string ``"pong"``.
+
+
+Streams
+-------
+
+Each stream is meant as a unidirectional communication channel, where the
+BigchainDB node is the only party sending messages (except for `keep-alive
+pings <#connection-keep-alive>`_). Any messages sent to the BigchainDB node
+(except the keep-alive pings) will be ignored.
+
+Streams will always be under the WebSocket protocol (so ``ws://`` or ``wss://``)
+and accessible as extensions to the ``/api/v<version>/streams/`` API root URL
+(for example, `validated transactions <#valid-transactions>`_ is accessible
+under ``/api/v1/streams/valid_tx``). If you're running your own BigchainDB
+instance and need help determining its root URL, you can look :ref:`here <determining-the-api-root-url>`.
+
+All messages sent in a stream are in the JSON format.
+
+.. note::
+
+    For simplicity, BigchainDB initially only provides a stream for all
+    validated transactions. In the future, we may provide streams for other
+    information, such as new blocks, new votes, or invalid transactions. We may
+    also provide the ability to filter the stream for specific qualities, such
+    as a specific ``output``'s ``public_key``.
+
+Valid Transactions
+~~~~~~~~~~~~~~~~~~
+
+``/valid_tx``
+
+Streams the IDs of any newly validated transactions. Message bodies will also
+contain the block ID containing the validated transaction.
+
+Example message::
+
+    {
+        "txid": "<sha3-256 hash>",
```

Comment:
> I'd prob include the asset_id in here too. Or just full transaction bodies. Otherwise people will have no way to know if they want to see the contents other than to do a GET /api/v1/transactions/....

+1 for `asset_id`. I'd like to have the full body as well but this might require having filtered event streams first.

> Changing two static strings (../transactions/ and ../blocks) of API routes that ideally do not change again from now on is definitely something we can afford.

I don't think it's a good idea. A transaction has an `id` that is it's hash. If we put the URL, then who is receiving events might need to "unpack" the URL to have the `id`. Also, a URL is tied to a host, and we have a federation of hosts, so which hosts would you use?